### PR TITLE
multiqc: possible fix for name collisions

### PR DESCRIPTION
**Pull request by matthdsm**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by ewels** [comment]
PR is based on discussions about generalising MultiQC inputs and my suggestion on syntax here: https://github.com/nextflow-io/nextflow/discussions/2842#discussioncomment-2681082

**Comment by mahesh-panchal** [comment]
Does the "??" limit it to 100 files?

**Comment by ewels** [comment]
> Does the "??" limit it to 100 files?

Possibly, not sure - may be possible to test fairly quickly with just `?`

If it does, then maybe `*/*` would work? Going from the docs? https://nextflow.io/docs/latest/process.html#multiple-input-files

**Comment by ewels** [comment]
Or rather `PREFIX*/*` - I have vague memory of playing with this years ago and I think you might need that text prefix in order to get the numbers.. 🤔

**Comment by Emiller88** [comment]
Looks good to me. Could we maybe make a 100 file test?

**Comment by matthdsm** [comment]
Using `*/*` also works.
How would you suggest simulating more name colissions?

**Comment by matthdsm** [comment]
this one creates a subdir per input:
```
├── 01
│   └── multiqc_config.yml -> /research/nextflow/nf-cmgg-preprocessing/assets/multiqc_config.yml
├── 02
│   └── workflow_summary_mqc.yaml -> /research/nextflow/nf-cmgg-preprocessing/work/tmp/1f/f7410ad0b8e205db20991d42675bf0/workflow_summary_mqc.yaml
├── 03
│   └── Reports -> /research/nextflow/nf-cmgg-preprocessing/work/d7/cfaaa668f0ec2cd1abce1bd9139131/Reports
├── 04
│   └── Sample1_S1_L001.fastp.json -> /research/nextflow/nf-cmgg-preprocessing/work/47/78311f05ef45a2daa1f318ac45dd95/Sample1_S1_L001.fastp.json
├── 05
│   └── Sample1.merged.metrics.txt -> /research/nextflow/nf-cmgg-preprocessing/work/73/cdc9a8916d98cd64796a4f9f1b09a2/Sample1.merged.metrics.txt
├── 06
│   └── Sample1.merged.bam.idxstats -> /research/nextflow/nf-cmgg-preprocessing/work/8a/aedd9c154ab9ed4fa5bc2b30eed3c9/Sample1.merged.bam.idxstats
├── 07
│   └── Sample1.merged.bam.stats -> /research/nextflow/nf-cmgg-preprocessing/work/02/29db81075fc47904392f4d9575fc6d/Sample1.merged.bam.stats
├── 08
│   └── Sample1.CollectMultipleMetrics.alignment_summary_metrics -> /research/nextflow/nf-cmgg-preprocessing/work/b0/be3523b958add52d8344fc35f9c339/Sample1.CollectMultipleMetrics.alignment_summary_metrics
├── 09
│   └── Sample1.CollectMultipleMetrics.base_distribution_by_cycle_metrics -> /research/nextflow/nf-cmgg-preprocessing/work/b0/be3523b958add52d8344fc35f9c339/Sample1.CollectMultipleMetrics.base_distribution_by_cycle_metrics
├── 10
│   └── Sample1.CollectMultipleMetrics.quality_by_cycle_metrics -> /research/nextflow/nf-cmgg-preprocessing/work/b0/be3523b958add52d8344fc35f9c339/Sample1.CollectMultipleMetrics.quality_by_cycle_metrics
├── 11
│   └── Sample1.CollectMultipleMetrics.quality_distribution_metrics -> /research/nextflow/nf-cmgg-preprocessing/work/b0/be3523b958add52d8344fc35f9c339/Sample1.CollectMultipleMetrics.quality_distribution_metrics
├── 12
│   └── Sample1.merged.bam.flagstat -> /research/nextflow/nf-cmgg-preprocessing/work/71/107c05f8e1a222b3a474a27c07d8a1/Sample1.merged.bam.flagstat
├── 13
│   └── software_versions_mqc.yml -> /research/nextflow/nf-cmgg-preprocessing/work/7f/3aca98a3e548f3da4e879d621250ee/software_versions_mqc.yml
├── multiqc_data
│   ├── bamsormadup_bamsormadup.txt
│   ├── multiqc.log
│   ├── multiqc_bclconvert_bylane.txt
│   ├── multiqc_bclconvert_bysample.txt
│   ├── multiqc_citations.txt
│   ├── multiqc_data.json
│   ├── multiqc_fastp.txt
│   ├── multiqc_general_stats.txt
│   ├── multiqc_picard_AlignmentSummaryMetrics.txt
│   ├── multiqc_picard_baseContent.txt
│   ├── multiqc_picard_quality_by_cycle.txt
│   ├── multiqc_picard_quality_score_distribution.txt
│   ├── multiqc_samtools_flagstat.txt
│   ├── multiqc_samtools_idxstats.txt
│   ├── multiqc_samtools_stats.txt
│   ├── multiqc_sources.txt
│   ├── picard_histogram.txt
│   ├── picard_histogram_1.txt
│   └── picard_histogram_2.txt
├── multiqc_report.html
└── versions.yml

File: tests/modules/multiqc/main.nf [hunk 1]
```diff
@@ -2,15 +2,30 @@
...
 
     FASTQC  ( input )
-    MULTIQC ( FASTQC.out.zip.collect { it[1] } )
+    MULTIQC ( FASTQC.out.zip.collect { it[1] }, [[],[]] )
+}
+
+workflow test_multiqc_fn_collision {
+    fqc_input = [ [ id: 'test', single_end: false ],
+                    [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)]
+                ]
```
Comment (mahesh-panchal) [inline tests/modules/multiqc/main.nf:21]:
```suggestion
    fqc_input = [ 
        [ id: 'test', single_end: false ],
        [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)]
    ]
```
Easier to see the elements of the input imo

File: modules/multiqc/main.nf [hunk 2]
```diff
@@ -20,8 +21,13 @@ process MULTIQC {
 
     script:
     def args = task.ext.args ?: ''
+    def config = multiqc_config ?: '--config $multiqc_config'
```
Comment (mahesh-panchal) [inline modules/multiqc/main.nf:24]:
```suggestion
    def config = multiqc_config ?: "--config $multiqc_config"
```
Should fix the test

File: modules/multiqc/main.nf [hunk 3]
```diff
@@ -20,8 +21,13 @@ process MULTIQC {
 
     script:
     def args = task.ext.args ?: ''
+    def config = multiqc_config ?: "--config $multiqc_config"
```
Comment (mahesh-panchal) [inline modules/multiqc/main.nf:24]:
```suggestion
    def config = multiqc_config ? "--config $multiqc_config" : ''
```
Only want the config when the value is set, not unset.

Comment (matthdsm) [reply modules/multiqc/main.nf:24]: Oh right, stupid mistake

File: modules/multiqc/main.nf [hunk 4]
```diff
@@ -20,8 +21,13 @@ process MULTIQC {
...
     script:
     def args = task.ext.args ?: ''
+    def config = multiqc_config ? "--config $multiqc_config" : ''
     """
-    multiqc -f $args .
+    multiqc \
+        --force \
+        $config \
+        $args \
+        .
```
Comment (mahesh-panchal) [inline modules/multiqc/main.nf:30]:
```suggestion
    multiqc \\
        --force \\
        $config \\
        $args \\
        .
```
Need double slashes here to have the new line in the shell script.

File: tests/modules/multiqc/main.nf [hunk 5]
```diff
@@ -2,15 +2,31 @@
...
-include { MULTIQC } from '../../../modules/multiqc/main.nf'
+include { FASTQC            } from '../../../modules/fastqc/main.nf'
+include { FASTQC as FASTQC2 } from '../../../modules/fastqc/main.nf'
+include { MULTIQC           } from '../../../modules/multiqc/main.nf'
 
 workflow test_multiqc {
     input = [ [ id: 'test', single_end: false ],
-              [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
-                file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+                [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)]
```
Comment (mahesh-panchal) [inline tests/modules/multiqc/main.nf:11]: Would you mind formatting this too please?

**Review by mahesh-panchal — APPROVED** [review]
Nice work

File: tests/modules/multiqc/main.nf (hunk 5 above)
Comment (matthdsm) [reply tests/modules/multiqc/main.nf]: done!

**Comment by matthdsm** [comment]
Thank you everyone!

**Pull request merged**